memory_reference/src/algo.hpp

```diff
@@ -0,0 +1,118 @@
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+#include "tags.hpp"
+#include "mem_ref_detail.hpp"
+#include "cupy_ref.hpp"
+#include "cupy_caster.hpp"
+#include "cupy_allocate.hpp"
+
+template<typename TDevice>
+class Algo {
+    public:
+        void whoami();
+
+        Algo();
+        void initialize_array(int size);
+        Mem_Ref<TDevice> get_input_memory();
+        void compute(Mem_Ref<TDevice> input, Mem_Ref<TDevice> output);
+        Mem_Ref<TDevice> get_ouput_memory();
+};
+
+template<>
+class Algo<CPU> {
+public:
+    Mem_Ref<CPU> input;
```

Comment:
Because for numpy the storage and reference object has the same type (`pybind11::array_t`) it is working. But please replace it with `pybind11::array_t`, because we want to keep the storage and not the reference. We want to create a reference, if it is needed (e.g. as return type). That both types are the same is a implementation detail, which should be not be visible for the user.